packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift

```diff
@@ -0,0 +1,257 @@
+  case idle
+  case checking
+  case downloading
+  case restarting
+}
+
+/**
+ All the possible types of events that can be sent to the machine. Each event
+ will cause the machine to transition to a new state.
+ */
+internal enum UpdatesStateEventType: String {
+  case check
+  case checkCompleteUnavailable
+  case checkCompleteAvailable
+  case checkError
+  case download
+  case downloadComplete
+  case downloadError
+  case restart
+}
+
+// MARK: - Data structures
+
+/**
+ Structure representing an event that can be sent to the machine.
+ Convenience getters are provided to get derived properties that will be
+ used to modify the context when the machine processes an event.
+ */
+internal struct UpdatesStateEvent {
+  var type: UpdatesStateEventType
+  var body: [String: Any] = [:]
+  var manifest: [String: Any]? {
+    guard let manifest = self.body["manifest"] as? [String: Any] else {
+      return nil
+    }
+    return manifest
+  }
+  var error: Error? {
+    guard let message = self.body["message"] as? String else {
+      return nil
+    }
+    return UpdatesStateException(message)
+  }
+  var isRollback: Bool {
+    guard let isRollback = self.body["isRollBackToEmbedded"] as? Bool else {
+      return false
+    }
+    return isRollback
+  }
+}
+
+/**
+ The state machine context, with information that will be readable from JS.
+ */
+internal struct UpdatesStateContext {
+  var isUpdateAvailable: Bool = false
+  var isUpdatePending: Bool = false
+  var isRollback: Bool = false
+  var isChecking: Bool = false
+  var isDownloading: Bool = false
+  var isRestarting: Bool = false
+  var latestManifest: [String: Any]?
+  var downloadedManifest: [String: Any]?
+  var checkError: Error?
+  var downloadError: Error?
+
+  var json: [String: Any] {
+    return [
+      "isUpdateAvailable": self.isUpdateAvailable,
+      "isUpdatePending": self.isUpdatePending,
+      "isRollback": self.isRollback,
+      "isChecking": self.isChecking,
+      "isDownloading": self.isDownloading,
+      "isRestarting": self.isRestarting,
+      // We pass NSNulls to keep this as a [String: Any] map
+      "latestManifest": self.latestManifest ?? NSNull(),
+      "downloadedManifest": self.downloadedManifest ?? NSNull(),
+      "checkError": self.checkError ?? NSNull(),
+      "downloadError": self.downloadError ?? NSNull()
+    ] as [String: Any]
+  }
+}
+
+// MARK: - State machine class
+
+/**
+ The Updates state machine class. There should be only one instance of this class
+ in a production app, instantiated as a property of AppController.
+ */
+internal class UpdatesStateMachine {
+  private let logger = UpdatesLogger()
+
+  init(changeEventDelegate: (any UpdatesStateChangeDelegate)) {
+    self.changeEventDelegate = changeEventDelegate
+  }
+
+  // MARK: - Public methods and properties
+
+  /**
+   In production, this is the AppController instance.
+   */
+  internal let changeEventDelegate: (any UpdatesStateChangeDelegate)
```

Comment:
Let's discuss this

Reply:
I was under the impression that all delegates needed to be weak to prevent retain cycles (controller retains state machine retains controller). I guess weak requires it be a var and be nullable, but that's technically the most correct since we want a tear down of the controller to not be blocked by this.

Reply:
At the moment, the state machine needs a delegate in order to be initialized, so the state machine gets created after the controller calls `super.init()` -- meaning the state machine property of the controller has to be nullable (weak). We can certainly do it the other way (state machine has nullable delegate, gets created by the controller before `super.init()`, then set the state machine's delegate property after).

Reply:
Not the nullability that I'm worried about. More the retain cycles. It's still fine that it needs a delegate to be initialized, but during deinit of an object, it needs to deference all strong references and it can't do that if the object being dereferenced has a strong reference to the owning object.

https://linear.app/expo/issue/ENG-9047/weak-delegate-in-updatesstatemachine